common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -531,6 +541,65 @@ protected SecretKey loadSecretKey(@NonNull final KeyType keyType) throws IOExcep
         throw new GeneralSecurityException(ErrorStrings.UNKNOWN_ERROR);
     }
 
+    /**
+     * If needed, get the key from inactive broker (if there is one).
+     * If it fails to get a new key. It will create a new one.
+     *
+     * This function can only be invoked by Broker apps.
+     */
+    public void migrateEncryptionKeyIfNeeded() throws GeneralSecurityException, IOException {
+        final String methodName = ":migrateEncryptionKeyIfNeeded";
+
+        final String activeBrokerPackageName = getPackageName();
+
+        if (!(COMPANY_PORTAL_APP_PACKAGE_NAME.equalsIgnoreCase(activeBrokerPackageName) ||
+                AZURE_AUTHENTICATOR_APP_PACKAGE_NAME.equalsIgnoreCase(activeBrokerPackageName))) {
+            final String errorMessage = "Caller is not a broker. Migration is not needed";
+            Logger.error(TAG + methodName, errorMessage, null);
+            throw new IllegalStateException(errorMessage);
+        }
+
+        mEncryptionKey = getKey();
+        if (mEncryptionKey != null){
+            Logger.verbose(TAG + methodName, "Key already exists.");
+            return;
+        }
+
+        mEncryptionKey = getKeyFromInactiveBroker(activeBrokerPackageName);
+        if (mEncryptionKey != null){
+            Logger.verbose(TAG + methodName, "Key is successfully retrieved from inactive broker. Saving key...");
+            saveSecretKey(mEncryptionKey);
+            return;
+        }
+
+        Logger.verbose(TAG + methodName, "Key migration failed. Create a new key.");
+        mEncryptionKey = createKey();
+        return;
+    }
+
+    @Nullable
+    protected SecretKey getKeyFromInactiveBroker(@NonNull final String activeBrokerPackageName) {
```

Comment:
>getKeyFromInactiveBroker [](start = 24, length = 24)

This naming is throwing my mind off - Are we supposed to get the key from the current inactive broker? If so, why is activeBrokerPackageName has getPackageName() output value (Line 553)?

Reply:
It takes activeBrokerPackageName to help determine what could possibly be the inactive broker.
i.e. if activeBrokerPackageName = CP, then inactive broker = authApp

Would adding more comment, or renaming help? Do you have any suggestion?

Reply:
removed